cloudmarker/alerts/emailalert.py

```diff
@@ -68,10 +77,44 @@ def done(self):
             self.body = ''.join(self.stringbuffer)
         message.attach(MIMEText(self.body))
 
-        smtp_connection = smtplib.SMTP(host=self.host, port=self.port)
+        smtp = self._prepare_smtp_session()
         try:
-            smtp_connection.sendmail(self.sender, self.to, message.as_string())
+            smtp.sendmail(self.sender, self.to, message.as_string())
         except smtplib.SMTPException as e:
             _log.error('Failed to send email: %s', e)
         finally:
-            smtp_connection.quit()
+            smtp.quit()
+
+    def _prepare_smtp_session(self):
+        """Return SMTP connection object.
+
+        Create a SMTP connection based on whether the SSL param.
+        If SSL connection is required (for eg: in case of gmail) then
+        method will return a SMTP_SSL connection. In other cases if the
+        authentication is not required or SMTP_SSL connection is not
+        required then a plain SMTP connection object is returned.
+        # For gmail follow the steps given below even to work for 2 factor
+        authentication
+        # 1. Log-in into Gmail with your account
+        # 2. Navigate to https://security.google.com/settings/security/
+        apppasswords
+        # 3. In 'select app' choose 'custom', give it an arbitrary name and
+        press generate
+        # 4. It will give you 16 chars token.
+        # 5. Use that token as password for login
+        # 6. Host is smtp.gmail.com and port is 465
+        """
+        # In case no authentication is required then no need to login
+        if self.use_ssl:
+            smtp = smtplib.SMTP_SSL(host=self.host, port=self.port)
+            smtp.set_debuglevel(True)
+            if self.username:
+                smtp.login(self.username, self.password)
+            return smtp
+
+        smtp = smtplib.SMTP(host=self.host, port=self.port)
+        smtp.set_debuglevel(True)
+        smtp.starttls()
+        if self.username:
+            smtp.login(self.username, self.password)
+        return smtp
```

Comment:
I see repetition of some statements twice. Unless there is a strong reason, repetition of code should be avoided because it becomes a maintenance overhead in future.

This code can certainly be written without repetition. One example would be:

```python
        if self.use_ssl:
            smtp = smtplib.SMTP_SSL(self.host, self.port)
        else:
            smtp = smtplib.SMTP(self.host, self.port)

        smtp.set_debuglevel(True)
        if not self.use_ssl:
            smtp.starttls()

        if self.username:
            smtp.login(self.username, self.password)

        return smtp
```

Another, perhaps slightly clearer way of writing this with an overhead of a repetition of one statement could be:

```python
        if self.use_ssl:
            smtp = smtplib.SMTP_SSL(self.host, self.port)
            smtp.set_debuglevel(True)
        else:
            smtp = smtplib.SMTP(self.host, self.port)
            smtp.set_debuglevel(True)
            smtp.starttls()

        if self.username:
            smtp.login(self.username, self.password)

        return smtp
```

Note: Please remember to prefix all internal use instance attributes with `_` (underscore) like I mentioned in another comment.

Reply:
https://github.com/cloudmarker/cloudmarker/pull/75/commits/f72e4709b765dccbf91e95c431acadb93ef30db1#diff-28d4faa43f2d620f5a2c54123370ee47R109